<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<title><code>too-many-cells</code> Workshop</title>
<meta name="generator" content="Org mode" />
<meta name="author" content="Gregory W. Schwartz" />
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .equation-container {
    display: table;
    text-align: center;
    width: 100%;
  }
  .equation {
    vertical-align: middle;
  }
  .equation-label {
    display: table-cell;
    text-align: right;
    vertical-align: middle;
  }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link rel="stylesheet" type="text/css" href="https://fniessen.github.io/org-html-themes/styles/readtheorg/css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="https://fniessen.github.io/org-html-themes/styles/readtheorg/css/readtheorg.css"/>
<script src="https://ajax.googleapis.com/ajax/libs/jquery/2.1.3/jquery.min.js"></script>
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/js/bootstrap.min.js"></script>
<script type="text/javascript" src="https://fniessen.github.io/org-html-themes/styles/lib/js/jquery.stickytableheaders.min.js"></script>
<script type="text/javascript" src="https://fniessen.github.io/org-html-themes/styles/readtheorg/js/readtheorg.js"></script>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2020 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">
<h1 class="title"><code>too-many-cells</code> Workshop</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#orga635606">Install <code>too-many-cells</code></a>
<ul>
<li><a href="#org1d791dd">Install <code>too-many-cells</code></a></li>
<li><a href="#org123793b">Adding to path (with stack installation)</a></li>
<li><a href="#org20717ba">Testing the installation</a></li>
</ul>
</li>
<li><a href="#org168212d">Data download</a>
<ul>
<li><a href="#org07f158d">Download brain data</a></li>
<li><a href="#org5ce1c74">Download heart data</a></li>
</ul>
</li>
<li><a href="#org69ea957">Prevent overlapping</a>
<ul>
<li><a href="#org200e464">Backup barcodes</a></li>
<li><a href="#org9e21fbe">Edit barcodes</a></li>
</ul>
</li>
<li><a href="#org0bc2e13">Tree creation with <code>too-many-cells</code></a>
<ul>
<li><a href="#org11459fd">Initial tree creation</a></li>
</ul>
</li>
<li><a href="#orgcb4c1ce">Coloring the <code>too-many-cells</code> tree.</a>
<ul>
<li><a href="#org20b6e19">Prepare labels file</a></li>
<li><a href="#org4018e3e">Color tree</a></li>
<li><a href="#org217c8f3">Custom colors</a></li>
</ul>
</li>
<li><a href="#org4c8127c">Getting more information from the tree</a>
<ul>
<li><a href="#orge31f4fa">Overlay modularity</a></li>
</ul>
</li>
<li><a href="#org58f9c8a">Pruning the tree</a>
<ul>
<li><a href="#org2f30bf6">Prune tree by size</a></li>
<li><a href="#orgf55f3f2">Prune tree by size distribution</a></li>
</ul>
</li>
<li><a href="#org527bd1a">Gene expression</a>
<ul>
<li><a href="#orgc9de737">Neuron marker overlay</a></li>
<li><a href="#org5a33b05">Increasing visibility</a></li>
<li><a href="#org06c54b3">Gene symbol rather than Ensembl</a></li>
<li><a href="#orgde43004">Heart marker overlay</a></li>
<li><a href="#org6b53483">Multiple gene expression overlays</a></li>
<li><a href="#org10f2c97">Custom colors for multiple expression overlays</a></li>
</ul>
</li>
<li><a href="#orgae73722">Differential expression</a>
<ul>
<li><a href="#orgfb948e9">Overlay node numbers</a></li>
<li><a href="#orgaf7312b">Differential expression for two nodes</a></li>
<li><a href="#orgf359839">Differential expression for two groups of nodes</a></li>
<li><a href="#org40d6a75">Label-filtered differential expression</a></li>
<li><a href="#org5f8b210">Gene distribution plots</a></li>
<li><a href="#orgec2e4cf">All to all!</a></li>
<li><a href="#orge1ade57">More to discover</a></li>
</ul>
</li>
</ul>
</div>
</div>
<p>
This is an instructional example of using <code>too-many-cells</code> meant to demonstrate
typical usage, originally presented in the IFI Advanced Computational Biology
Club.
</p>

<p>
For more information about <code>too-many-cells</code>:
</p>

<p>
<a href="https://gregoryschwartz.github.io/too-many-cells/">Website</a>
</p>

<p>
See <a href="https://github.com/GregorySchwartz/too-many-cells">https://github.com/GregorySchwartz/too-many-cells</a> for latest version.
</p>

<p>
See <a href="https://doi.org/10.1038/s41592-020-0748-5">the publication</a> (and please cite!) for more information about the algorithm.
</p>

<div id="outline-container-orga635606" class="outline-2">
<h2 id="orga635606">Install <code>too-many-cells</code></h2>
<div class="outline-text-2" id="text-orga635606">
</div>
<div id="outline-container-org1d791dd" class="outline-3">
<h3 id="org1d791dd">Install <code>too-many-cells</code></h3>
<div class="outline-text-3" id="text-org1d791dd">
<p>
Follow instructions on <a href="https://gregoryschwartz.github.io/too-many-cells/">https://gregoryschwartz.github.io/too-many-cells/</a> for
details. First, clone the <code>too-many-cells</code> repository.
</p>

<div class="org-src-container">
<pre class="src src-sh">git clone https://github.com/GregorySchwartz/too-many-cells.git
</pre>
</div>

<p>
Enter the folder and install with <a href="https://nixos.org/nix/">nix</a>.
</p>

<div class="org-src-container">
<pre class="src src-sh"><span style="color: #ff8700;">cd</span> ./too-many-cells
nix-env -f default.nix -i too-many-cells
</pre>
</div>
</div>
</div>

<div id="outline-container-org123793b" class="outline-3">
<h3 id="org123793b">Adding to path (with stack installation)</h3>
<div class="outline-text-3" id="text-org123793b">
<p>
If using stack, the resulting binary will install to <code>~/.local/bin</code>. Add to
<code>$PATH</code> so you can invoke the command from anywhere!
</p>

<div class="org-src-container">
<pre class="src src-sh"><span style="color: #ff8700;">export</span> <span style="color: #87afaf;">PATH</span>=$<span style="color: #87afaf;">HOME</span>/.local/bin:$<span style="color: #87afaf;">PATH</span>
</pre>
</div>

<p>
This command will only work in the current shell. To permanently add to path,
add the previous line to <code>~/.bashrc</code> or <code>~/.profile</code>.
</p>
</div>
</div>

<div id="outline-container-org20717ba" class="outline-3">
<h3 id="org20717ba">Testing the installation</h3>
<div class="outline-text-3" id="text-org20717ba">
<p>
Test to see if the installation worked when in path:
</p>

<div class="org-src-container">
<pre class="src src-sh">too-many-cells -h
</pre>
</div>

<pre class="example">
too-many-cells, Gregory W. Schwartz. Clusters and analyzes single cell data.

Usage: too-many-cells (make-tree | interactive | differential | diversity |
                      paths | classify | peaks | motifs | matrix-output)

Available options:
  -h,--help                Show this help text

Available commands:
  make-tree                
  interactive              
  differential             
  diversity                
  paths                    
  classify                 
  peaks                    
  motifs                   
  matrix-output 
</pre>
</div>
</div>
</div>

<div id="outline-container-org168212d" class="outline-2">
<h2 id="org168212d">Data download</h2>
<div class="outline-text-2" id="text-org168212d">
</div>
<div id="outline-container-org07f158d" class="outline-3">
<h3 id="org07f158d">Download brain data</h3>
<div class="outline-text-3" id="text-org07f158d">
<p>
We'll need data from 10x. Let's cluster mouse brain and heart cells from
<a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/neuron_1k_v3">https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/neuron_1k_v3</a>
and
<a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/heart_1k_v3">https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/heart_1k_v3</a>
as a quick, illustrative example here. <b>Note</b>: these are modern formats of
<code>cellranger</code> outputs (<code>v3</code>), but <code>too-many-cells</code> works with both older and
newer formats.
</p>

<div class="org-src-container">
<pre class="src src-shell"><span style="color: #767676;"># </span><span style="color: #767676;">Make the data directory</span>
mkdir -p data/brain

<span style="color: #767676;"># </span><span style="color: #767676;">Enter the directory</span>
<span style="color: #ff8700;">cd</span> ./data/brain

<span style="color: #767676;"># </span><span style="color: #767676;">Download the data</span>
wget http://cf.10xgenomics.com/samples/cell-exp/3.0.0/neuron_1k_v3/neuron_1k_v3_filtered_feature_bc_matrix.tar.gz

<span style="color: #767676;"># </span><span style="color: #767676;">Uncompress data</span>
tar xvf neuron_1k_v3_filtered_feature_bc_matrix.tar.gz
</pre>
</div>

<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">


<colgroup>
<col  class="org-left" />
</colgroup>
<tbody>
<tr>
<td class="org-left">filtered_feature_bc_matrix/</td>
</tr>

<tr>
<td class="org-left">filtered_feature_bc_matrix/features.tsv.gz</td>
</tr>

<tr>
<td class="org-left">filtered_feature_bc_matrix/matrix.mtx.gz</td>
</tr>

<tr>
<td class="org-left">filtered_feature_bc_matrix/barcodes.tsv.gz</td>
</tr>
</tbody>
</table>
</div>
</div>

<div id="outline-container-org5ce1c74" class="outline-3">
<h3 id="org5ce1c74">Download heart data</h3>
<div class="outline-text-3" id="text-org5ce1c74">
<p>
Let's do the same for the heart cells:
</p>

<div class="org-src-container">
<pre class="src src-shell"><span style="color: #767676;"># </span><span style="color: #767676;">Make the data directory</span>
mkdir -p data/heart

<span style="color: #767676;"># </span><span style="color: #767676;">Enter the directory</span>
<span style="color: #ff8700;">cd</span> ./data/heart

<span style="color: #767676;"># </span><span style="color: #767676;">Download the data</span>
wget http://cf.10xgenomics.com/samples/cell-exp/3.0.0/heart_1k_v3/heart_1k_v3_filtered_feature_bc_matrix.tar.gz

<span style="color: #767676;"># </span><span style="color: #767676;">Uncompress data</span>
tar xvf heart_1k_v3_filtered_feature_bc_matrix.tar.gz
</pre>
</div>

<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">


<colgroup>
<col  class="org-left" />
</colgroup>
<tbody>
<tr>
<td class="org-left">filtered_feature_bc_matrix/</td>
</tr>

<tr>
<td class="org-left">filtered_feature_bc_matrix/features.tsv.gz</td>
</tr>

<tr>
<td class="org-left">filtered_feature_bc_matrix/barcodes.tsv.gz</td>
</tr>

<tr>
<td class="org-left">filtered_feature_bc_matrix/matrix.mtx.gz</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>

<div id="outline-container-org69ea957" class="outline-2">
<h2 id="org69ea957">Prevent overlapping</h2>
<div class="outline-text-2" id="text-org69ea957">
</div>
<div id="outline-container-org200e464" class="outline-3">
<h3 id="org200e464">Backup barcodes</h3>
<div class="outline-text-3" id="text-org200e464">
<p>
These matrices both use <code>BARCODE-1</code> as their cell identifiers. If aggregating
with <code>cellranger</code> this won't be an issue, but because we aren't doing that let's
make sure there are no conflicts. First, let's backup our barcodes as we will be
making changes to ensure no overlapping.
</p>

<div class="org-src-container">
<pre class="src src-shell">cp ./data/brain/filtered_feature_bc_matrix/barcodes.tsv<span style="color: #008787;">{</span>.gz,.gz.bk<span style="color: #008787;">}</span>
cp ./data/heart/filtered_feature_bc_matrix/barcodes.tsv<span style="color: #008787;">{</span>.gz,.gz.bk<span style="color: #008787;">}</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org9e21fbe" class="outline-3">
<h3 id="org9e21fbe">Edit barcodes</h3>
<div class="outline-text-3" id="text-org9e21fbe">
<p>
Now let's edit the heart barcodes to have <code>-2</code> instead of <code>-1</code>.
</p>

<div class="org-src-container">
<pre class="src src-shell">cat ./data/heart/filtered_feature_bc_matrix/barcodes.tsv.gz.bk <span style="color: #afaf00;">\</span>
  | gzip -d <span style="color: #afaf00;">\</span>
  | sed <span style="color: #afaf00;">"s/-1/-2/g"</span> <span style="color: #afaf00;">\</span>
  | gzip <span style="color: #afaf00;">\</span>
  &gt; ./data/heart/filtered_feature_bc_matrix/barcodes.tsv.gz

cat ./data/heart/filtered_feature_bc_matrix/barcodes.tsv.gz | gzip -d | head
</pre>
</div>

<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">


<colgroup>
<col  class="org-left" />
</colgroup>
<tbody>
<tr>
<td class="org-left">AAACCCACACCAGTAT-2</td>
</tr>

<tr>
<td class="org-left">AAACCCAGTCACCTTC-2</td>
</tr>

<tr>
<td class="org-left">AAACCCAGTGGAACAC-2</td>
</tr>

<tr>
<td class="org-left">AAACGAAAGTGCCCGT-2</td>
</tr>

<tr>
<td class="org-left">AAACGAAGTCAGGTGA-2</td>
</tr>

<tr>
<td class="org-left">AAAGGATAGCACCGAA-2</td>
</tr>

<tr>
<td class="org-left">AAAGGATGTAACGGTG-2</td>
</tr>

<tr>
<td class="org-left">AAAGGGCAGGACGGAG-2</td>
</tr>

<tr>
<td class="org-left">AAAGTGACAGAACATA-2</td>
</tr>

<tr>
<td class="org-left">AAAGTGATCAAAGGAT-2</td>
</tr>
</tbody>
</table>

<p>
That's it! This will help when we assign labels to each cell later on.
</p>
</div>
</div>
</div>

<div id="outline-container-org0bc2e13" class="outline-2">
<h2 id="org0bc2e13">Tree creation with <code>too-many-cells</code></h2>
<div class="outline-text-2" id="text-org0bc2e13">
</div>
<div id="outline-container-org11459fd" class="outline-3">
<h3 id="org11459fd">Initial tree creation</h3>
<div class="outline-text-3" id="text-org11459fd">
<p>
We now have everything we need for initial runs with <code>too-many-cells</code>! Let's begin
by building a tree (ignore <code>printf</code> throughout this document, they are just
reporting the resulting file). We can specify multiple matrices to combine
automatically.
</p>

<div class="org-src-container">
<pre class="src src-shell">too-many-cells make-tree <span style="color: #afaf00;">\</span>
  --matrix-path ./data/brain/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --matrix-path ./data/heart/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --filter-thresholds <span style="color: #afaf00;">"(250, 1)"</span> <span style="color: #afaf00;">\</span>
  --output out <span style="color: #afaf00;">\</span>
  &gt; clusters.csv

<span style="color: #ff8700;">printf</span> <span style="color: #afaf00;">"./out/dendrogram.svg"</span>
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="./out/dendrogram.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>

<p>
The initial tree is built! It tells us the tree structure and the number of
cells in each leaf. Want to actually see which cells are brain and which cells
are heart? Let's give the tree some colors!
</p>
</div>
</div>
</div>

<div id="outline-container-orgcb4c1ce" class="outline-2">
<h2 id="orgcb4c1ce">Coloring the <code>too-many-cells</code> tree.</h2>
<div class="outline-text-2" id="text-orgcb4c1ce">
</div>
<div id="outline-container-org20b6e19" class="outline-3">
<h3 id="org20b6e19">Prepare labels file</h3>
<div class="outline-text-3" id="text-org20b6e19">
<p>
We can color the tree using any label. In this case, we want to give each cell a
label based on it's celltype from the data set. Let's quickly do that.
</p>

<div class="org-src-container">
<pre class="src src-shell">gzip -d -c ./data/brain/filtered_feature_bc_matrix/barcodes.tsv.gz ./data/heart/filtered_feature_bc_matrix/barcodes.tsv.gz <span style="color: #afaf00;">\</span>
  | sed <span style="color: #afaf00;">"s/-1/-1,Brain/"</span> <span style="color: #afaf00;">\</span>
  | sed <span style="color: #afaf00;">"s/-2/-2,Heart/"</span> <span style="color: #afaf00;">\</span>
  | sed <span style="color: #afaf00;">"1i item,label"</span> <span style="color: #afaf00;">\</span>
  &gt; labels.csv

head ./labels.csv
</pre>
</div>

<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">


<colgroup>
<col  class="org-left" />

<col  class="org-left" />
</colgroup>
<tbody>
<tr>
<td class="org-left">item</td>
<td class="org-left">label</td>
</tr>

<tr>
<td class="org-left">AAACGAATCAAAGCCT-1</td>
<td class="org-left">Brain</td>
</tr>

<tr>
<td class="org-left">AAACGCTGTAATGTGA-1</td>
<td class="org-left">Brain</td>
</tr>

<tr>
<td class="org-left">AAACGCTGTCCTGGGT-1</td>
<td class="org-left">Brain</td>
</tr>

<tr>
<td class="org-left">AAAGAACCAGGACATG-1</td>
<td class="org-left">Brain</td>
</tr>

<tr>
<td class="org-left">AAAGGTACACACGGTC-1</td>
<td class="org-left">Brain</td>
</tr>

<tr>
<td class="org-left">AAAGTCCAGTCACTAC-1</td>
<td class="org-left">Brain</td>
</tr>

<tr>
<td class="org-left">AAAGTCCGTGACTGTT-1</td>
<td class="org-left">Brain</td>
</tr>

<tr>
<td class="org-left">AAAGTCCTCCAGCCTT-1</td>
<td class="org-left">Brain</td>
</tr>

<tr>
<td class="org-left">AAAGTGAGTTCCTAAG-1</td>
<td class="org-left">Brain</td>
</tr>
</tbody>
</table>
</div>
</div>

<div id="outline-container-org4018e3e" class="outline-3">
<h3 id="org4018e3e">Color tree</h3>
<div class="outline-text-3" id="text-org4018e3e">
<p>
Great! Now we just need to feed it to <code>too-many-cells</code>. <b>Note</b>: We use <code>--prior</code>
from now on so we don't need to calculate the tree all over again. This argument
makes things much faster!
</p>

<div class="org-src-container">
<pre class="src src-shell">too-many-cells make-tree <span style="color: #afaf00;">\</span>
  --prior out <span style="color: #afaf00;">\</span>
  --labels-file ./labels.csv <span style="color: #afaf00;">\</span>
  --dendrogram-output <span style="color: #afaf00;">"tree_labeled.svg"</span> <span style="color: #afaf00;">\</span>
  --output out <span style="color: #afaf00;">\</span>
  &gt; clusters.csv

<span style="color: #ff8700;">printf</span> <span style="color: #afaf00;">"./out/tree_labeled.svg"</span>
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="./out/tree_labeled.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>
</div>
</div>

<div id="outline-container-org217c8f3" class="outline-3">
<h3 id="org217c8f3">Custom colors</h3>
<div class="outline-text-3" id="text-org217c8f3">
<p>
We can also change the colors however we want:
</p>

<div class="org-src-container">
<pre class="src src-shell">too-many-cells make-tree <span style="color: #afaf00;">\</span>
  --prior out <span style="color: #afaf00;">\</span>
  --labels-file ./labels.csv <span style="color: #afaf00;">\</span>
  --draw-colors <span style="color: #afaf00;">"[\"#66c2a5\", \"#fc8d62\"]"</span> <span style="color: #afaf00;">\</span>
  --dendrogram-output <span style="color: #afaf00;">"tree_labeled_alternate.svg"</span> <span style="color: #afaf00;">\</span>
  --output out <span style="color: #afaf00;">\</span>
  &gt; clusters.csv

<span style="color: #ff8700;">printf</span> <span style="color: #afaf00;">"./out/tree_labeled_alternate.svg"</span>
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="./out/tree_labeled_alternate.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>
</div>
</div>
</div>

<div id="outline-container-org4c8127c" class="outline-2">
<h2 id="org4c8127c">Getting more information from the tree</h2>
<div class="outline-text-2" id="text-org4c8127c">
</div>
<div id="outline-container-orge31f4fa" class="outline-3">
<h3 id="orge31f4fa">Overlay modularity</h3>
<div class="outline-text-3" id="text-orge31f4fa">
<p>
Now that we have a basic tree, we can start doing some quick edits. Want
modularity overlays to show the modularity at each node in the tree?
</p>

<div class="org-src-container">
<pre class="src src-shell">too-many-cells make-tree <span style="color: #afaf00;">\</span>
  --prior out <span style="color: #afaf00;">\</span>
  --labels-file ./labels.csv <span style="color: #afaf00;">\</span>
  --dendrogram-output <span style="color: #afaf00;">"tree_modularity.svg"</span> <span style="color: #afaf00;">\</span>
  --draw-mark <span style="color: #afaf00;">"MarkModularity"</span> <span style="color: #afaf00;">\</span>
  --output out <span style="color: #afaf00;">\</span>
  &gt; clusters.csv

<span style="color: #ff8700;">printf</span> <span style="color: #afaf00;">"./out/tree_modularity.svg"</span>
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="./out/tree_modularity.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>
</div>
</div>
</div>

<div id="outline-container-org58f9c8a" class="outline-2">
<h2 id="org58f9c8a">Pruning the tree</h2>
<div class="outline-text-2" id="text-org58f9c8a">
</div>
<div id="outline-container-org2f30bf6" class="outline-3">
<h3 id="org2f30bf6">Prune tree by size</h3>
<div class="outline-text-3" id="text-org2f30bf6">
<p>
For a large number of cells, the tree can grow quite large. To prune the tree,
we can use different cutoffs. However, this will change the tree structure, so
be sure to output the tree in a different folder to avoid overwriting the
original tree (so we can still use <code>--prior</code>)! Let's have no leaf with less than
30 cells:
</p>

<div class="org-src-container">
<pre class="src src-shell">too-many-cells make-tree <span style="color: #afaf00;">\</span>
  --prior out <span style="color: #afaf00;">\</span>
  --labels-file ./labels.csv <span style="color: #afaf00;">\</span>
  --min-size <span style="color: #d787af;">30</span> <span style="color: #afaf00;">\</span>
  --output out_pruned <span style="color: #afaf00;">\</span>
  &gt; clusters_pruned.csv

<span style="color: #ff8700;">printf</span> <span style="color: #afaf00;">"./out_pruned/dendrogram.svg"</span>
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="./out_pruned/dendrogram.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>
</div>
</div>

<div id="outline-container-orgf55f3f2" class="outline-3">
<h3 id="orgf55f3f2">Prune tree by size distribution</h3>
<div class="outline-text-3" id="text-orgf55f3f2">
<p>
Don't want arbitrary number cutoffs? Let's instead make a cutoff using the
distribution of cluster sizes. We can use <code>--smart-cutoff</code> to look at the
distribution of cluster sizes, split proportions, or distances (modularity here)
and select certain median absolute deviations (MADs) away from the median as a
cutoff. We select which feature to create a distribution by using that feature's
normal cutoff argument, where the cutoff value is ignored (so we can put 1, for
instance, as it will be ignored). Let's revise our previous attempt by cutting 1
MAD away from the median node size:
</p>

<div class="org-src-container">
<pre class="src src-shell">too-many-cells make-tree <span style="color: #afaf00;">\</span>
  --prior out <span style="color: #afaf00;">\</span>
  --labels-file ./labels.csv <span style="color: #afaf00;">\</span>
  --min-size <span style="color: #d787af;">1</span> <span style="color: #afaf00;">\</span>
  --smart-cutoff <span style="color: #d787af;">1</span> <span style="color: #afaf00;">\</span>
  --dendrogram-output <span style="color: #afaf00;">"tree_smart.svg"</span> <span style="color: #afaf00;">\</span>
  --output out_pruned <span style="color: #afaf00;">\</span>
  &gt; clusters_pruned.csv

<span style="color: #ff8700;">printf</span> <span style="color: #afaf00;">"./out_pruned/tree_smart.svg"</span>
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="./out_pruned/tree_smart.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>
</div>
</div>
</div>

<div id="outline-container-org527bd1a" class="outline-2">
<h2 id="org527bd1a">Gene expression</h2>
<div class="outline-text-2" id="text-org527bd1a">
</div>
<div id="outline-container-orgc9de737" class="outline-3">
<h3 id="orgc9de737">Neuron marker overlay</h3>
<div class="outline-text-3" id="text-orgc9de737">
<p>
Want to overlay gene expression? We'll need the matrices again, but still use
<code>--prior</code> to avoid clustering. Also, we use normalization to avoid looking at
only the counts, but rather normalized counts. Let's look neuron cell marker
<i>Rbfox3</i>:
</p>

<div class="org-src-container">
<pre class="src src-shell">too-many-cells make-tree <span style="color: #afaf00;">\</span>
  --matrix-path ./data/brain/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --matrix-path ./data/heart/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --filter-thresholds <span style="color: #afaf00;">"(250, 1)"</span> <span style="color: #afaf00;">\</span>
  --prior out <span style="color: #afaf00;">\</span>
  --normalization <span style="color: #afaf00;">"UQNorm"</span> <span style="color: #afaf00;">\</span>
  --draw-leaf <span style="color: #afaf00;">"DrawItem (DrawContinuous [\"ENSMUSG00000025576\"])"</span> <span style="color: #afaf00;">\</span>
  --dendrogram-output <span style="color: #afaf00;">"tree_neuron.svg"</span> <span style="color: #afaf00;">\</span>
  --output out <span style="color: #afaf00;">\</span>
  &gt; clusters.csv

<span style="color: #ff8700;">printf</span> <span style="color: #afaf00;">"./out/tree_neuron.svg"</span>
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="./out/tree_neuron.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>
</div>
</div>

<div id="outline-container-org5a33b05" class="outline-3">
<h3 id="org5a33b05">Increasing visibility</h3>
<div class="outline-text-3" id="text-org5a33b05">
<p>
Can't see too well? Let's up the saturation!
</p>

<div class="org-src-container">
<pre class="src src-shell">too-many-cells make-tree <span style="color: #afaf00;">\</span>
  --matrix-path ./data/brain/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --matrix-path ./data/heart/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --filter-thresholds <span style="color: #afaf00;">"(250, 1)"</span> <span style="color: #afaf00;">\</span>
  --prior out <span style="color: #afaf00;">\</span>
  --normalization <span style="color: #afaf00;">"UQNorm"</span> <span style="color: #afaf00;">\</span>
  --draw-leaf <span style="color: #afaf00;">"DrawItem (DrawContinuous [\"ENSMUSG00000025576\"])"</span> <span style="color: #afaf00;">\</span>
  --dendrogram-output <span style="color: #afaf00;">"tree_neuron_saturated.svg"</span> <span style="color: #afaf00;">\</span>
  --draw-scale-saturation <span style="color: #d787af;">6</span> <span style="color: #afaf00;">\</span>
  --output out <span style="color: #afaf00;">\</span>
  &gt; clusters.csv

<span style="color: #ff8700;">printf</span> <span style="color: #afaf00;">"./out/tree_neuron_saturated.svg"</span>
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="./out/tree_neuron_saturated.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>
</div>
</div>

<div id="outline-container-org06c54b3" class="outline-3">
<h3 id="org06c54b3">Gene symbol rather than Ensembl</h3>
<div class="outline-text-3" id="text-org06c54b3">
<p>
Want to use the gene symbol? <code>cellranger</code> provides that! Let's use that feature
column:
</p>

<div class="org-src-container">
<pre class="src src-shell">too-many-cells make-tree <span style="color: #afaf00;">\</span>
  --matrix-path ./data/brain/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --matrix-path ./data/heart/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --filter-thresholds <span style="color: #afaf00;">"(250, 1)"</span> <span style="color: #afaf00;">\</span>
  --prior out <span style="color: #afaf00;">\</span>
  --feature-column <span style="color: #d787af;">2</span> <span style="color: #afaf00;">\</span>
  --normalization <span style="color: #afaf00;">"UQNorm"</span> <span style="color: #afaf00;">\</span>
  --draw-leaf <span style="color: #afaf00;">"DrawItem (DrawContinuous [\"Rbfox3\"])"</span> <span style="color: #afaf00;">\</span>
  --dendrogram-output <span style="color: #afaf00;">"tree_neuron_gene_symbol.svg"</span> <span style="color: #afaf00;">\</span>
  --draw-scale-saturation <span style="color: #d787af;">6</span> <span style="color: #afaf00;">\</span>
  --output out <span style="color: #afaf00;">\</span>
  &gt; clusters.csv

<span style="color: #ff8700;">printf</span> <span style="color: #afaf00;">"./out/tree_neuron_gene_symbol.svg"</span>
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="./out/tree_neuron_gene_symbol.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>
</div>
</div>

<div id="outline-container-orgde43004" class="outline-3">
<h3 id="orgde43004">Heart marker overlay</h3>
<div class="outline-text-3" id="text-orgde43004">
<p>
What about heart?
</p>

<div class="org-src-container">
<pre class="src src-shell">too-many-cells make-tree <span style="color: #afaf00;">\</span>
  --matrix-path ./data/brain/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --matrix-path ./data/heart/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --filter-thresholds <span style="color: #afaf00;">"(250, 1)"</span> <span style="color: #afaf00;">\</span>
  --prior out <span style="color: #afaf00;">\</span>
  --feature-column <span style="color: #d787af;">2</span> <span style="color: #afaf00;">\</span>
  --normalization <span style="color: #afaf00;">"UQNorm"</span> <span style="color: #afaf00;">\</span>
  --draw-leaf <span style="color: #afaf00;">"DrawItem (DrawContinuous [\"Gata6\"])"</span> <span style="color: #afaf00;">\</span>
  --dendrogram-output <span style="color: #afaf00;">"tree_heart.svg"</span> <span style="color: #afaf00;">\</span>
  --draw-scale-saturation <span style="color: #d787af;">6</span> <span style="color: #afaf00;">\</span>
  --output out <span style="color: #afaf00;">\</span>
  &gt; clusters.csv

<span style="color: #ff8700;">printf</span> <span style="color: #afaf00;">"./out/tree_heart.svg"</span>
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="./out/tree_heart.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>
</div>
</div>

<div id="outline-container-org6b53483" class="outline-3">
<h3 id="org6b53483">Multiple gene expression overlays</h3>
<div class="outline-text-3" id="text-org6b53483">
<p>
What about <b>both</b>!?
</p>

<div class="org-src-container">
<pre class="src src-shell">too-many-cells make-tree <span style="color: #afaf00;">\</span>
  --matrix-path ./data/brain/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --matrix-path ./data/heart/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --filter-thresholds <span style="color: #afaf00;">"(250, 1)"</span> <span style="color: #afaf00;">\</span>
  --prior out <span style="color: #afaf00;">\</span>
  --feature-column <span style="color: #d787af;">2</span> <span style="color: #afaf00;">\</span>
  --normalization <span style="color: #afaf00;">"UQNorm"</span> <span style="color: #afaf00;">\</span>
  --draw-leaf <span style="color: #afaf00;">"DrawItem (DrawThresholdContinuous [(\"Rbfox3\", Exact 0), (\"Gata6\", Exact 0)])"</span> <span style="color: #afaf00;">\</span>
  --dendrogram-output <span style="color: #afaf00;">"tree_brain_heart_markers.svg"</span> <span style="color: #afaf00;">\</span>
  --output out <span style="color: #afaf00;">\</span>
  &gt; clusters.csv

<span style="color: #ff8700;">printf</span> <span style="color: #afaf00;">"./out/tree_brain_heart_markers.svg"</span>
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="./out/tree_brain_heart_markers.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>
</div>
</div>

<div id="outline-container-org10f2c97" class="outline-3">
<h3 id="org10f2c97">Custom colors for multiple expression overlays</h3>
<div class="outline-text-3" id="text-org10f2c97">
<p>
Now we can see a more complete picture! But this is combinatorial in the number
of features with high and low, can we focus in on a few? The order is always
alphabetical, so we can assign our own colors &#x2013; ones that don't get saturated
for unimportant cases!
</p>

<div class="org-src-container">
<pre class="src src-shell">too-many-cells make-tree <span style="color: #afaf00;">\</span>
  --matrix-path ./data/brain/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --matrix-path ./data/heart/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --filter-thresholds <span style="color: #afaf00;">"(250, 1)"</span> <span style="color: #afaf00;">\</span>
  --prior out <span style="color: #afaf00;">\</span>
  --feature-column <span style="color: #d787af;">2</span> <span style="color: #afaf00;">\</span>
  --normalization <span style="color: #afaf00;">"UQNorm"</span> <span style="color: #afaf00;">\</span>
  --draw-leaf <span style="color: #afaf00;">"DrawItem (DrawThresholdContinuous [(\"Rbfox3\", Exact 0), (\"Gata6\", Exact 0)])"</span> <span style="color: #afaf00;">\</span>
  --dendrogram-output <span style="color: #afaf00;">"tree_brain_heart_markers_alternate.svg"</span> <span style="color: #afaf00;">\</span>
  --draw-colors <span style="color: #afaf00;">"[\"#e41a1c\", \"#377eb8\", \"#4daf4a\", \"#999999\"]"</span> <span style="color: #afaf00;">\</span>
  --draw-scale-saturation <span style="color: #d787af;">6</span> <span style="color: #afaf00;">\</span>
  --output out <span style="color: #afaf00;">\</span>
  &gt; clusters.csv

<span style="color: #ff8700;">printf</span> <span style="color: #afaf00;">"./out/tree_brain_heart_markers_alternate.svg"</span>
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="./out/tree_brain_heart_markers_alternate.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>
</div>
</div>
</div>

<div id="outline-container-orgae73722" class="outline-2">
<h2 id="orgae73722">Differential expression</h2>
<div class="outline-text-2" id="text-orgae73722">
</div>
<div id="outline-container-orgfb948e9" class="outline-3">
<h3 id="orgfb948e9">Overlay node numbers</h3>
<div class="outline-text-3" id="text-orgfb948e9">
<p>
Now that we've seen some expressions, we quantify the differences in
expressions between populations using the <code>differential</code> entry point. First,
which node is which? Let's overlay their node IDs.
</p>

<div class="org-src-container">
<pre class="src src-shell">too-many-cells make-tree <span style="color: #afaf00;">\</span>
  --prior out <span style="color: #afaf00;">\</span>
  --labels-file ./labels.csv <span style="color: #afaf00;">\</span>
  --draw-node-number <span style="color: #afaf00;">\</span>
  --dendrogram-output <span style="color: #afaf00;">"tree_numbers.svg"</span> <span style="color: #afaf00;">\</span>
  --output out <span style="color: #afaf00;">\</span>
  &gt; clusters.csv

<span style="color: #ff8700;">printf</span> <span style="color: #afaf00;">"./out/tree_numbers.svg"</span>
</pre>
</div>


<div class="figure">
<p><object type="image/svg+xml" data="./out/tree_numbers.svg" class="org-svg">
Sorry, your browser does not support SVG.</object>
</p>
</div>
</div>
</div>

<div id="outline-container-orgaf7312b" class="outline-3">
<h3 id="orgaf7312b">Differential expression for two nodes</h3>
<div class="outline-text-3" id="text-orgaf7312b">
<p>
We know that <code>Gata6</code> is higher in node 94 versus 121. Let's look at the
differential expression of 94 / 121.
</p>

<div class="org-src-container">
<pre class="src src-shell">too-many-cells differential <span style="color: #afaf00;">\</span>
  --matrix-path ./data/brain/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --matrix-path ./data/heart/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --filter-thresholds <span style="color: #afaf00;">"(250, 1)"</span> <span style="color: #afaf00;">\</span>
  --prior out <span style="color: #afaf00;">\</span>
  --nodes <span style="color: #afaf00;">"([121], [94])"</span> <span style="color: #afaf00;">\</span>
  &gt; ./out/94_vs_121.csv

<span style="color: #ff8700;">printf</span> <span style="color: #afaf00;">"./out/94_vs_121.csv"</span>
</pre>
</div>

<p>
<a href="./out/94_vs_121.csv">./out/94_vs_121.csv</a>
</p>
</div>
</div>

<div id="outline-container-orgf359839" class="outline-3">
<h3 id="orgf359839">Differential expression for two groups of nodes</h3>
<div class="outline-text-3" id="text-orgf359839">
<p>
Why does the format have brackets? Because it's a list! We can compare
multiple nodes to each other:
</p>

<div class="org-src-container">
<pre class="src src-shell">too-many-cells differential <span style="color: #afaf00;">\</span>
  --matrix-path ./data/brain/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --matrix-path ./data/heart/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --filter-thresholds <span style="color: #afaf00;">"(250, 1)"</span> <span style="color: #afaf00;">\</span>
  --prior out <span style="color: #afaf00;">\</span>
  --feature-column <span style="color: #d787af;">2</span> <span style="color: #afaf00;">\</span>
  --normalization <span style="color: #afaf00;">"UQNorm"</span> <span style="color: #afaf00;">\</span>
  --nodes <span style="color: #afaf00;">"([121], [94, 5])"</span> <span style="color: #afaf00;">\</span>
  &gt; ./out/94_5_vs_121.csv

<span style="color: #ff8700;">printf</span> <span style="color: #afaf00;">"./out/94_5_vs_121.csv"</span>
</pre>
</div>

<p>
<a href="./out/94_5_vs_121.csv">./out/94_5_vs_121.csv</a>
</p>
</div>
</div>

<div id="outline-container-org40d6a75" class="outline-3">
<h3 id="org40d6a75">Label-filtered differential expression</h3>
<div class="outline-text-3" id="text-org40d6a75">
<p>
There's some nodes with multiple celltypes in them &#x2013; can we compare just
the brain cells to heart cells in their "respective" cluster?
</p>

<div class="org-src-container">
<pre class="src src-shell">too-many-cells differential <span style="color: #afaf00;">\</span>
  --matrix-path ./data/brain/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --matrix-path ./data/heart/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --filter-thresholds <span style="color: #afaf00;">"(250, 1)"</span> <span style="color: #afaf00;">\</span>
  --prior out <span style="color: #afaf00;">\</span>
  --labels-file ./labels.csv <span style="color: #afaf00;">\</span>
  --feature-column <span style="color: #d787af;">2</span> <span style="color: #afaf00;">\</span>
  --normalization <span style="color: #afaf00;">"UQNorm"</span> <span style="color: #afaf00;">\</span>
  --nodes <span style="color: #afaf00;">"([121], [94, 5])"</span> <span style="color: #afaf00;">\</span>
  --labels <span style="color: #afaf00;">"([\"Brain\"], [\"Heart\"])"</span> <span style="color: #afaf00;">\</span>
  &gt; ./out/94_5_vs_121_filtered.csv

<span style="color: #ff8700;">printf</span> <span style="color: #afaf00;">"./out/94_5_vs_121_filtered.csv"</span>
</pre>
</div>

<p>
<a href="./out/94_5_vs_121_filtered.csv">./out/94_5_vs_121_filtered.csv</a>
</p>
</div>
</div>

<div id="outline-container-org5f8b210" class="outline-3">
<h3 id="org5f8b210">Gene distribution plots</h3>
<div class="outline-text-3" id="text-org5f8b210">
<p>
We can do some basic plotting as well for specific genes, such as <code>Gata6</code> and
<code>Rbfox3</code> here.
</p>

<div class="org-src-container">
<pre class="src src-shell">too-many-cells differential <span style="color: #afaf00;">\</span>
  --matrix-path ./data/brain/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --matrix-path ./data/heart/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --filter-thresholds <span style="color: #afaf00;">"(250, 1)"</span> <span style="color: #afaf00;">\</span>
  --labels-file ./labels.csv <span style="color: #afaf00;">\</span>
  --prior out <span style="color: #afaf00;">\</span>
  --feature-column <span style="color: #d787af;">2</span> <span style="color: #afaf00;">\</span>
  --normalization <span style="color: #afaf00;">"UQNorm"</span> <span style="color: #afaf00;">\</span>
  --nodes <span style="color: #afaf00;">"([121], [94, 5])"</span> <span style="color: #afaf00;">\</span>
  --labels <span style="color: #afaf00;">"([\"Brain\"], [\"Heart\"])"</span> <span style="color: #afaf00;">\</span>
  --features <span style="color: #afaf00;">"Gata6"</span> <span style="color: #afaf00;">\</span>
  --features <span style="color: #afaf00;">"Rbfox3"</span> <span style="color: #afaf00;">\</span>
  --plot-output <span style="color: #afaf00;">"./out/genes.pdf"</span>

<span style="color: #ff8700;">printf</span> <span style="color: #afaf00;">"./out/genes.pdf"</span>
</pre>
</div>

<p>
<a href="./out/genes.pdf">./out/genes.pdf</a>
</p>
</div>
</div>

<div id="outline-container-orgec2e4cf" class="outline-3">
<h3 id="orgec2e4cf">All to all!</h3>
<div class="outline-text-3" id="text-orgec2e4cf">
<p>
Last, but not least, we can get the differential genes for every node versus all
other nodes by not specifying any nodes at all. Useful for quick gene enrichment
analyses!
</p>

<div class="org-src-container">
<pre class="src src-shell">too-many-cells differential <span style="color: #afaf00;">\</span>
  --matrix-path ./data/brain/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --matrix-path ./data/heart/filtered_feature_bc_matrix/ <span style="color: #afaf00;">\</span>
  --filter-thresholds <span style="color: #afaf00;">"(250, 1)"</span> <span style="color: #afaf00;">\</span>
  --prior out <span style="color: #afaf00;">\</span>
  --feature-column <span style="color: #d787af;">2</span> <span style="color: #afaf00;">\</span>
  --normalization <span style="color: #afaf00;">"UQNorm"</span> <span style="color: #afaf00;">\</span>
  --nodes <span style="color: #afaf00;">"([], [])"</span> <span style="color: #afaf00;">\</span>
  &gt; ./out/all_nodes_differential.csv

<span style="color: #ff8700;">printf</span> <span style="color: #afaf00;">"./out/all_nodes_differential.csv"</span>
</pre>
</div>

<p>
<a href="./out/all_nodes_differential.csv">./out/all_nodes_differential.csv</a>
</p>
</div>
</div>

<div id="outline-container-orge1ade57" class="outline-3">
<h3 id="orge1ade57">More to discover</h3>
<div class="outline-text-3" id="text-orge1ade57">
<p>
We've only scratched the surface here, for many more customizable options, check
out the help documentation for each entry point, e.g. <code>too-many-cells make-tree
-h</code>.
</p>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Gregory W. Schwartz</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
